client/app/widgets/QuickCreate/__tests__/QuickCreate.spec.jsx

```diff
@@ -0,0 +1,179 @@
+      />,
+    );
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+  describe('When the component renders', () => {
+    it('renders an input with a label with the accordion closed', () => {
+      expect(screen.queryByTestId('accordion-closed')).toBeInTheDocument();
+      expect(getNodeText(screen.getByText('label'))).toEqual('label');
+    });
+  });
+  describe('when accordion is clicked', () => {
+    it('it opens and the user can interact with the textbox', () => {
+      expect(screen.queryByTestId('accordion-closed')).toBeInTheDocument();
+      const quickCreateLabel = screen.getByText('label');
+      // user clicks on the label to open accordion
+      userEvent.click(quickCreateLabel);
+      // accordion reflects open state
+      expect(screen.queryByTestId('accordion-open')).toBeInTheDocument();
+      const userInput = 'my input text';
+      // User enters some text
+      userEvent.type(screen.getByRole('textbox'), userInput);
+      // Closes the input
+      userEvent.click(quickCreateLabel);
+      // Reopen
+      userEvent.click(quickCreateLabel);
+      expect(screen.getByRole('textbox')).toHaveValue(userInput);
+    });
+  });
+  describe('when input is changed', () => {
+    it('opens modal when the checkbox does not already exist', () => {
+      // open accordion
+      userEvent.click(screen.getByText('label'));
+      // modal doesn't exist yet
+      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
+      // user enters a checkbox value that does not exist
+      userEvent.type(screen.getByRole('textbox'), 'new checkbox{enter}');
+      // modal appears
+      expect(screen.queryByRole('dialog')).toBeInTheDocument();
+    });
+    it('does not open the modal if the checkbox already exists', () => {
+      // open accordion
+      userEvent.click(screen.getByText('label'));
+      // type a value that already exists
+      userEvent.type(
+        screen.getByRole('textbox'),
+        `${checkboxes[0].label}{enter}`,
+      );
+      // modal should not be open
+      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
+    });
+  });
+
+  describe('when the form', () => {
+    it('is submitted it adds checkboxes from data', async () => {
+      const response = {
+        data: {
+          success: true,
+          id: 1,
+          name: 'new checkbox',
+          slug: 'new-checkbox',
+        },
+      };
+      const axiosPostSpy = jest
+        .spyOn(axios, 'post')
+        .mockImplementation(() => Promise.resolve(response));
+      // open accordion
+      userEvent.click(screen.getByText('label'));
+      // enter a value that doesnt exist
+      userEvent.type(screen.getByRole('textbox'), 'new checkbox{enter}');
+      // submit the form
+      userEvent.click(screen.getByRole('button', { name: 'Submit' }));
+      // when the dialog has been removed there was a successful submission
+      await waitForElementToBeRemoved(() => screen.getByRole('dialog'));
```

Comment:
I was seeing the act warning and not being familiar with it, did some investigating. The current approach is from https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning . First video is the one I got it from, around the 5:20 mark is the final code. 

Using waitForElementToBeRemove() moves the testing code closer to the user experience because from the user POV they click submit and they wait for some interaction. In our case thats the dialog being removed. 

The above suggestion is also more implementation rather than behavior testing. However, there doesn't seem to be a strong opinion in the rtl docs.

Reply:
That is true. Checking first that the view has responded makes sense for an integration test. Seeing the spy being tested also, it was naturally tempting to think of it in sequence, but that isn’t necessary 👍 it works either way, but waiting for the element first seems natural.